src/v/http/request_builder.h

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "bytes/iobuf.h"
+#include "thirdparty/ada/ada.h"
+
+#include <seastar/core/sstring.hh>
+
+#include <absl/container/flat_hash_map.h>
+#include <absl/container/flat_hash_set.h>
+#include <boost/beast/http/message.hpp>
+
+namespace http {
+
+// Builds a request using the builder pattern. Allows setting the host, target
+// (path), method, headers and query params.
+class request_builder {
+public:
+    using error_type = ss::sstring;
+    static constexpr auto default_state{"host not set"};
+
+    using expected
+      = tl::expected<boost::beast::http::request_header<>, error_type>;
+
+    // The host supplied here is parsed and stored as a result. When the request
+    // is finally built this is added as a host header. If the parse failed then
+    // build returns an error variant.
+    request_builder& host(std::string_view host);
+
+    request_builder& target(std::string_view target);
```

Comment:
I don't really see `target` here, why not use `path`? At the very least there should be docs.

Reply:
I added a comment to the method declaration. 

The `target` terminology for consistency with the underlying `boost::beast::http::request...` constructs which use the term target, it is also commonly used in cloud storage code, but since this is outside of cloud storage and path is more widely used generally with http requests, I can change it if it sounds more readable.

Reply:
I would personally prefer path over target, but it’s fine to keep if that is consistent with existing code.

Reply:
sure, changed to `path`